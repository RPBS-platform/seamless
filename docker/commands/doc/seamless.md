% seamless(1) Version 0.1 | seamless-cli Documentation

NAME
====

**seamless** - Starts a new Seamless Docker container and runs a command inside it.

SYNOPSIS
========

| **seamless** \[_command_]

DESCRIPTION
===========

Starts a new Seamless Docker container and runs a command inside it.

The current directory is mounted to /cwd, and the command is executed there

/tmp is mounted as well

Uses the host network for the Docker container. Will only work under Linux.

BUGS
====

See GitHub Issues: <https://github.com/sjdv1982/seamless/issues>

AUTHOR
======

Sjoerd de Vries <sjdv1982@gmail.com>

SEE ALSO
========

**seamless-bash(1)**, **seamless-ipython(1)**, **seamless-jupyter(1)**