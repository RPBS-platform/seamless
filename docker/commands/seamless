set -u -e

firstarg=$1 #there must be at least one argument
gdbflags='--cap-add=SYS_PTRACE --security-opt seccomp=unconfined --security-opt apparmor=unconfined'
docker run --rm \
  $gdbflags \
  --network=host \
  -v `pwd`:/cwd \
  -v /tmp:/tmp \
  --workdir /cwd \
  --user $(id -u):$(id -g) \
  rpbs/seamless $*
