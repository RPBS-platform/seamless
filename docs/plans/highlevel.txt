Status of the high level:
- Most things are now working. Macro is still missing, node wrapping is very incomplete,
    and too easy to construct an illegal graph, but mostly fine now.
- hcell, hrc and htf refer to the node. Should be renamed.
- Translation is now lazy, i.e. only upon translate(), equilibrate()
- node["TEMP"] refers to a literal (i.e. Python) construction value. Low-level cells are needed to
  parse this to a serializable value. Right now, the mid-level does this on the first translation pass, then pops the TEMP.
  This is very ugly. mid-level should be configured to invoke remote parsing, and remote parsing should be invokable without
  translation. This way, the translate.py can be disconnected, and the remaining mid-and-high-level routines could run in
  the browser (under Pyodide; Transcrypt is probably less useful since we would want a Python REPL). If *all* of the state
  is in the graph (i.e. Python Context, Cell etc are just thin wrappers), then the graph could be manipulated in parallel
  by JS, e.g. manipulating the position of the node on a visualization canvas).
  Right now this is not exactly the case: there is context._graph but ALSO context._children.
