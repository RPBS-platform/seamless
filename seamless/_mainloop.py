"""
Seamless mainloop routines

Status:
For now, the mainloop is controlled by Qt timers
An overhaul to put it under asyncio control is most welcome.
"""

import time
from collections import deque
import threading
import asyncio
import contextlib

FAILSAFE_WORK_LATENCY = 50 #latency of run_work in ms

_work = deque()
_priority_work = deque()
def add_work(work, priority=False):
    if priority:
        _priority_work.append(work)
    else:
        _work.append(work)

_running_work = False
_signal_processing = 0
def run_work():
    global _running_work
    if threading.current_thread() is not threading.main_thread():
        return
    if _running_work:
        return
    _running_work = True
    #print("WORKING", len(_priority_work), len(_work))
    #work_count = 0
    works = (_priority_work, _work)
    if _signal_processing > 0:
        works = (_priority_work,)
    for w in works:
        while len(w):
            work = w.popleft()
            try:
                work()
                #work_count += 1
            except Exception:
                traceback.print_exc()
            #if work_count == 100 and not _signal_processing:
            #    run_qt() # Necessary to prevent freezes in glwindow
            #    work_count = 0

    #Whenever work is done, do an asyncio flush
    loop = asyncio.get_event_loop()
    loop.call_soon(lambda loop: loop.stop(), loop)
    loop.run_forever()

    if _signal_processing == 0 and _run_qt:
        run_qt() # Necessary to prevent freezes in glwindow
    _running_work = False

def asyncio_finish():
    try:
        loop = asyncio.get_event_loop()
        loop.stop()
        loop.run_forever()
    except RuntimeError:
        pass

_run_qt = True  #set by __init__.py
event_loop = None #set by __init__.py

_qt_is_running = False
def run_qt():
    global _qt_is_running
    if _qt_is_running:
        return
    #Whenever work is done, let Qt flush its event queue
    # If you don't, segfaults happen (see test-gl-BUG.py)
    _qt_is_running = True
    event_loop.processEvents()
    _qt_is_running = False

def mainloop():
    """Only run in non-IPython mode"""
    while 1:
        run_work()
        time.sleep(FAILSAFE_WORK_LATENCY/1000)
