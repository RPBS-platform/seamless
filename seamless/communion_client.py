import asyncio

REMOTE_TIMEOUT = 5.0

class CommunionClient:
    pass

class CommunionBufferClient(CommunionClient):
    config_type = "buffer"

    def __init__(self, servant, config):
        self.servant = servant
        self.config_buffer = config["buffer"]
        self.config_status = config["buffer_status"]

    async def status(self, checksum):
        assert checksum is not None
        if not self.config_status:
            return
        message = {
            "type": "buffer_status",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        if result is not None:
            try:
                status, result_checksum = result
                if status in (0, 1):
                    communion_client_manager.remote_checksum_available.add(
                        result_checksum
                    )
            except:
                pass
        return result
    
    async def submit(self, checksum):
        assert checksum is not None
        if not self.config_buffer:
            return
        message = {
            "type": "buffer",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result


class CommunionBufferLengthClient(CommunionClient):
    config_type = "buffer_length"

    def __init__(self, servant, config):
        self.servant = servant

    async def submit(self, checksum):
        assert checksum is not None
        message = {
            "type": "buffer_length",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result

class CommunionTransformationClient(CommunionClient):
    config_type = "transformation"

    def __init__(self, servant, config):
        self.servant = servant
        self.config_job = config["transformation_job"]
        self.config_status = config["transformation_status"]

    async def status(self, checksum):
        assert checksum is not None
        if not self.config_status:
            return
        message = {
            "type": "transformation_status",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result
    
    async def submit(self, checksum):
        if not self.config_job:
            return
        message = {
            "type": "transformation_job",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result


class CommunionBuildModuleClient(CommunionClient):
    config_type = "build_module"

    def __init__(self, servant, config):
        self.servant = servant
        self.config_job = config["build_module_job"]
        self.config_status = config["build_module_status"]

    async def status(self, checksum):
        assert checksum is not None
        if not self.config_status:
            return
        message = {
            "type": "build_module_status",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result
    
    async def submit(self, checksum):
        if not self.config_job:
            return
        message = {
            "type": "transformation_job",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result

"""
class CommunionBuildModuleClient(CommunionClient):
    config_type = "build_module"
    cache_task_servers = remote_build_model_servers
    def _prepare_message(self, content):
        return {
            "type": self.config_type,
            "content": content,
        }

class CommunionTransformerResultClient(CommunionClient):
    config_type = "transformer_result"
    cache_task_servers = remote_transformer_result_servers
    def _prepare_message(self, checksum):
        return {
            "type": self.config_type,
            "content": checksum,
        }


class CommunionTransformerJobClient(CommunionPairClient):
    config_type = "transformer_job"
    cache_task_servers = remote_job_servers
    def _prepare_message(self, funcmode, checksum):
        if funcmode == "check":
            return {
                "type": "transformer_job_check",
                "content": checksum,
            }
        elif funcmode == "run":
            return {
                "type": "transformer_job_run",
                "content": checksum,
            }


"""

class CommunionClientManager:
    _clientclasses = [klass for klass in globals().values() if isinstance(klass, type) \
      and issubclass(klass, CommunionClient) and klass is not CommunionClient]

    _clientclasses = {klass.config_type:klass for klass in _clientclasses}

    def __init__(self):
        self.clients = {k:[] for k in self._clientclasses}
        self.remote_checksum_available = set()
        self.servant_to_clients = {}
        self.servant_to_peer_id = {}

    def add_servant(self, servant, peer_id, *,
            config_servant, config_master
    ):
        servid = id(servant)
        self.servant_to_clients[servid] = []
        self.servant_to_peer_id[servid] = peer_id
        communion_types = {
            "buffer": ["buffer", "buffer_status"],
            "buffer_length": ["buffer_length"],
            "transformation": ["transformation_job", "transformation_status"],
            "build_module": ["build_module_job", "build_module_status"],
        }
        for communion_type in communion_types:
            sub_communion_types = {}
            for sub_communion_type in communion_types[communion_type]:
                c_master = config_master[sub_communion_type]
                c_servant = config_servant[sub_communion_type]                
                if not c_master or not c_servant:
                    c = False
                elif sub_communion_type == "buffer":
                    if c_servant == "small" or c_master == "small":
                        c = "small"
                    else:
                        c = True
                else:
                    c = True
                sub_communion_types[sub_communion_type] = c
            if not any(sub_communion_types.values()):
                continue
            print("ADD SERVANT", communion_type)
            clientclass = self._clientclasses[communion_type]
            client = clientclass(servant, sub_communion_types)
            self.clients[communion_type].append(client)
            self.servant_to_clients[servid].append((communion_type, client))

    def remove_servant(self, servant):
        self.remote_checksum_available.clear()
        clients = self.servant_to_clients.pop(id(servant))
        self.servant_to_peer_id.pop(id(servant))
        for communion_type, client in clients:
            print("REMOVE SERVANT", communion_type)
            self.clients[communion_type].remove(client)

    async def remote_buffer_status(self, checksum, peer_id):
        if checksum in self.remote_checksum_available:
            return True
        clients = []
        for client in self.clients["buffer"]:
            client_peer_id = self.servant_to_peer_id[id(client.servant)]
            if client_peer_id != peer_id:
                clients.append(client)
        if not len(clients):
            return False
        coros = [client.status(checksum) for client in clients]
        futures = [asyncio.ensure_future(coro) for coro in coros]
        while 1:
            done, pending = await asyncio.wait(*futures, 
                timeout=REMOTE_TIMEOUT,
                return_when=asyncio.FIRST_COMPLETED
            )
            for future in done:
                try:
                    result = future.result()
                    if result > 0:
                        return True
                except:
                    continue
            if not len(pending):
                return False         


communion_client_manager = CommunionClientManager()
from .communion_server import communion_server