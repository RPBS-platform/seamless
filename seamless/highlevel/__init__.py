import inspect
from copy import deepcopy
from types import LambdaType
from ast import PyCF_ONLY_AST, FunctionDef, Expr, Lambda

from ..mixed import MixedBase
from ..silk import Silk
from ..silk.validation import _allowed_types
from ..core.lambdacode import lambdacode
from ..core.cached_compile import cached_compile

ConstantTypes = _allowed_types + (Silk, MixedBase, tuple)

def set_hcell(cell, value):
    from ..core.structured_cell import StructuredCellState
    if cell["celltype"] == "structured":
        cell["stored_state"] = StructuredCellState.from_data(value)
    else:
        cell["stored_value"] = value

def parse_function_code(code_or_func, identifier="<None>"):
    if callable(code_or_func):
        func = code_or_func
        code = inspect.getsource(func)
        if isinstance(func, LambdaType) and func.__name__ == "<lambda>":
            code = lambdacode(func)
            if code is None:
                raise ValueError("Cannot extract source code from this lambda")
    else:
        assert isinstance(code_or_func, str)
        code = code_or_func

    ast = cached_compile(code, identifier, "exec", PyCF_ONLY_AST)
    is_function = (len(ast.body) == 1 and
                   isinstance(ast.body[0], FunctionDef))

    if is_function:
        func_name = ast.body[0].name
        code_object = cached_compile(code, identifier, "exec")
    else:
        assert (len(ast.body) == 1 and isinstance(ast.body[0], Expr))
        assert isinstance(ast.body[0].value, Lambda)
        func_name = "<lambda>"
        code_object = cached_compile(code, identifier, "eval")
    return code, func_name, code_object

from .Context import Context
from .Library import stdlib, mylib
from .Reactor import Reactor
from .Transformer import Transformer
from .Cell import Cell
from .Link import Link
