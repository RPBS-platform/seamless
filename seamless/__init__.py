"""
Seamless: framework for data-driven and live programming
Copyright 2016-2018, Sjoerd de Vries
"""

import sys
import time
import functools
import traceback
import atexit

from abc import abstractmethod
class Wrapper:
    @abstractmethod
    def _unwrap(self):
        pass

from . import silk
from . import mixed

#Dependencies of seamless

# 1. hard dependencies; without these, "import seamless" will fail.
# Still, if necessary, some of these dependencies could be removed, but seamless would have to be more minimalist in loading its lib

import numpy as np

if np.dtype(np.object).itemsize != 8:
    raise ImportError("Seamless requires a 64-bit system")

#silk must be imported before mixed
from . import silk
from . import mixed

from .core import mainloop as _mainloop
from .core.mainloop import mainloop, mainloop_one_iteration, asyncio_finish, workqueue
atexit.register(asyncio_finish)

ipython = None
ipy_error = None
try:
    from IPython import get_ipython
    from IPython.core.error import UsageError
    from IPython.terminal.pt_inputhooks import register as _register_integration_terminal
    from ipykernel.eventloops import register_integration as _register_integration_kernel
except ImportError:
    raise
    ipy_error = "Cannot find IPython"
else:
    ipython = get_ipython()
    if ipython is None:
        ipy_error = "Seamless was not imported inside IPython"

if ipy_error is None:
    last_exception = None
    def new_except_hook(etype, evalue, tb):
        global last_exception
        exc = traceback.format_exception(etype, evalue, tb)
        if exc != last_exception:
            last_exception = exc
            print("".join(exc))

    def patch_excepthook():
        sys.excepthook = new_except_hook
    patch_excepthook()

    def mainloop():
        raise RuntimeError("Cannot run seamless.mainloop() in IPython mode")

    def inputhook_terminal(context):
        while not context.input_is_ready():
            mainloop_one_iteration()
    _register_integration_terminal("seamless", inputhook_terminal)

    @_register_integration_kernel('seamless')
    def inputhook_kernel(kernel):
        while 1:
            t = time.time()
            while time.time() - t < kernel._poll_interval:
                mainloop_one_iteration()        
            kernel.do_one_iteration()

    _register_integration_terminal("seamless", inputhook_terminal)
    
    ipython.enable_gui("seamless")
    
else:
    sys.stderr.write("    " + ipy_error + "\n")
    sys.stderr.write("    Call seamless.mainloop(), seamless.flush() or context.equilibrate() to process cell updates\n")

def flush():
    from .core.mainloop import workqueue
    workqueue.flush()

from .highlevel import *
from .silk import Silk
from .debugger import pdb
