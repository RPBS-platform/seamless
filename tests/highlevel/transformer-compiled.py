import os, tempfile
from seamless.highlevel import Context, Cell

ctx = Context()
ctx.mount(os.path.join(tempfile.gettempdir(), "transformer-compiled"))

ctx.transform = lambda a,b: a + b
ctx.transform.example.a = 0
ctx.transform.example.b = 0
ctx.result = ctx.transform
ctx.result.celltype = "json"
ctx.equilibrate()
print(ctx.result.value)

ctx.transform.language = "cpp"
ctx.code >> ctx.transform.code
ctx.code = """
extern "C" double transform(int a, int b) {
    return a + b;
}"""
ctx.transform.result.example = 0.0 #example, just to fill the schema
ctx.equilibrate()
print(ctx.result.value)

ctx.a = 10
ctx.a.celltype = "json"
ctx.transform.a = ctx.a

ctx.b = 30
ctx.b.celltype = "json"
ctx.transform.b = ctx.b

ctx.equilibrate()
print(ctx.result.value)
print(ctx.status())

"""
### The code generator itself  (gen_header, translator)
 can be hacked, but in only one way:

    from seamless.highlevel import stdlib
    t = stdlib.compiled_transformer
    t.gen_header.code.mount("/tmp/gen-header.py")

This will affect all transformers.
But: by default, "t.register_library()" must be invoked upon every change.
This is a design decision that can be overruled by specifying:
    t.auto_register(True)

See the documentation of highlevel/Library.py for more details.
"""

ctx.transform.debug = True
