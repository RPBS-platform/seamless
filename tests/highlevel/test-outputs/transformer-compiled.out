5

g++ -c main.cpp -O3 -ffast-math -march=native -fPIC -p -g -o main.o



g++ -c main.cpp -O3 -ffast-math -march=native -fPIC -p -g -o main.o


5.0

g++ -c main.cpp -O3 -ffast-math -march=native -fPIC -p -g -o main.o



g++ -c main.cpp -O3 -ffast-math -march=native -fPIC -p -g -o main.o


40.0
Status: OK
